File: Arrays/Problems/linear_search.cpp
#include <iostream>

using namespace std;

/* Time Complexity: O(n) i.e. Linear Time Complexity */
int linear_search(int array[], int size, int search)
{

    for (int i = 0; i < size; i++)
    {
        if (search == array[i])
        {
            return i;
        }
    }

    return -1;
}

int main()
{
    // Initial Array
    int array[] = {45, 67, 39, 23, 78, 34, 56};
    int size = sizeof(array) / sizeof(array[0]);
    int search;

    cout << "Enter your search: ";
    cin >> search;

    int index = linear_search(array, size, search);

    cout << "Element Found at index: " << index << endl;

    return 0;
}

File: Searching/linear_search.py
"""
Linear Search quick revision notes:
- Scan elements from left to right and compare each with the target.
- Works on unsorted data.
- Returns the first match when duplicates exist.
- Time: O(n) worst/average, O(1) best (target at first index).
- Space: O(1) auxiliary.
"""


def linear_search(elements, target):
    """Return index of first match, else False if target is not present."""
    for i in range(len(elements)):
        if target == elements[i]:
            return i

    return False


if __name__ == "__main__":
    elements = [7, 5, 4, 32, 9, 5, -1]
    target = 32
    index = linear_search(elements, target)

    # Use explicit check so index 0 is treated as a valid match.
    if index is not False:
        print(f"Element '{target}' found at index {index}")
    else:
        print("Not Found")
